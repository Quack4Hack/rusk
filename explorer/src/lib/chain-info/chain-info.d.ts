type BlockHeader = {
  date: Date;
  feespaid: number;
  hash: string;
  height: number;
  nextblockhash: string;
  prevblockhash: string;
  reward: number;
  seed: string;
  statehash: string;
};

type BlockTransactions = {
  data: Transaction[];
  stats: {
    averageGasPrice: number;
    gasLimit: number;
    gasUsed: number;
  };
};

type Block = {
  header: BlockHeader;
  transactions: BlockTransactions;
};

type ChainInfo = {
  blocks: Block[];
  transactions: Transaction[];
};

type SearchResult = {
  id: string;
  type: "block" | "transaction";
};

type Transaction = {
  blockhash: string;
  blockheight: number;
  contract: string;
  date: Date;
  feepaid: number;
  gaslimit: number;
  gasprice: number;
  gasspent: number;
  method: string;
  success: boolean;
  txerror: string;
  txid: string;
};
