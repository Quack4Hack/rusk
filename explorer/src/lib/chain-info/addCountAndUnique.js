/**
 * Returns an array of unique objects, sorted by the frequency of their occurrence,
 * with an additional count property representing the number of times
 * that city-country combination appeared in the input data.
 *
 * @param {Omit<NodeLocation, "count">[]} data
 * @returns {NodeLocation[]}
 */
function addCountAndUnique(data) {
  const countMap = data.reduce((locationMap, location) => {
    const key = `${location.city},${location.country}`;

    if (!locationMap[key]) {
      locationMap[key] = { ...location, count: 0 };
    }
    locationMap[key].count += 1;
    return locationMap;
  }, /** @type {{ [key: string]: NodeLocation }} */ ({}));

  // Return only unique city-country combinations
  return Object.values(countMap).sort((a, b) => b.count - a.count);
}

export default addCountAndUnique;
