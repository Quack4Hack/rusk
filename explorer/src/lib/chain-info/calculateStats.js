import {
  always,
  compose,
  condition,
  filterWith,
  getKey,
  partition,
} from "lamb";

import { arraySumByKey } from "$lib/dusk/array";

/** @type {(txs: Pick<GQLTransaction, "err">[]) => number} */
const getFailedTxAmount = compose(
  getKey("length"),
  filterWith((tx) => tx.err !== null)
);

/**
 * We take into account only provisioners with
 * the minimum stake amount of 1000 Dusk (1e12 Lux).
 *
 * @type {(provisioners: HostProvisioner[]) => HostProvisioner[]}
 */
const getValidProvisioners = filterWith((p) => p.amount >= 1e12);

/**
 * Sums the values of the "amount" key in the received array.
 * Returns zero if the array is empty.
 */
const sumByAmount = condition(
  (provisioners) => provisioners.length > 0,
  arraySumByKey("amount"),
  always(0)
);

/**
 * @param {HostProvisioner[]} provisioners
 * @param {number} lastHeight
 * @param {Pick<GQLTransaction, "err">[]} last100BlocksTxs
 * @returns {Stats}
 */
function calculateStats(provisioners, lastHeight, last100BlocksTxs) {
  const [activeProvisioners, waitingProvisioners] = partition(
    getValidProvisioners(provisioners),
    (p) => p.eligibility <= lastHeight
  );

  return {
    activeProvisioners: activeProvisioners.length,
    activeStake: sumByAmount(activeProvisioners),
    lastBlock: lastHeight,
    txs100blocks: {
      failed: getFailedTxAmount(last100BlocksTxs),
      transfers: last100BlocksTxs.length,
    },
    waitingProvisioners: waitingProvisioners.length,
    waitingStake: sumByAmount(waitingProvisioners),
  };
}

export default calculateStats;
