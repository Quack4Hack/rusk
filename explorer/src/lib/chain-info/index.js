export { default as transformBlock } from "./transformBlock";
export { default as transformTransaction } from "./transformTransaction";
