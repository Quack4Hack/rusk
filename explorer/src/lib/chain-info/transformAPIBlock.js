import { mapWith, pipe, skip, updateIn } from "lamb";

import { unixTsToDate } from "$lib/dusk/date";

import { transformAPITransaction } from ".";

/** @type {(v: APIBlockHeader) => Required<APIBlockHeader>} */
const mergeWithDefaults = (v) => ({
  nextblockhash: "",
  prevblockhash: "",
  statehash: "",
  ...v,
});

/** @type {(header: Required<APIBlockHeader>) => BlockHeader} */
const addHeaderDate = (header) => ({
  ...header,
  date: unixTsToDate(header.ts),
});

/** @type {(v: APIBlockHeader) => BlockHeader} */
const transformAPIBlockHeader = pipe([
  mergeWithDefaults,
  addHeaderDate,
  skip(["__typename", "timestamp", "ts", "version"]),
]);

/** @type {(v: APIBlock) => Block} */
const transformAPIBlock = ({ header, transactions }) => ({
  header: transformAPIBlockHeader(header),
  transactions: updateIn(
    transactions,
    "data",
    mapWith(transformAPITransaction)
  ),
});

export default transformAPIBlock;
