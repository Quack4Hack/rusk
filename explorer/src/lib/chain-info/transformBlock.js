import { mapWith, pipe, skip, updateIn } from "lamb";

import { unixTsToDate } from "$lib/dusk/date";

import { transformTransaction } from ".";

/** @type {(v: APIBlockHeader) => Required<APIBlockHeader>} */
const mergeWithDefaults = (v) => ({
  nextblockhash: "",
  prevblockhash: "",
  statehash: "",
  ...v,
});

/** @type {(header: Required<APIBlockHeader>) => BlockHeader} */
const addHeaderDate = (header) => ({
  ...header,
  date: unixTsToDate(header.ts),
});

/** @type {(v: APIBlockHeader) => BlockHeader} */
const transformBlockHeader = pipe([
  mergeWithDefaults,
  addHeaderDate,
  skip(["__typename", "timestamp", "ts"]),
]);

/** @type {(v: APIBlock) => Block} */
const transformBlock = ({ header, transactions }) => ({
  header: transformBlockHeader(header),
  transactions: updateIn(transactions, "data", mapWith(transformTransaction)),
});

export default transformBlock;
