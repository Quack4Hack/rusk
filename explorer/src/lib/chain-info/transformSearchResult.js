import { compose, getPathIn, mapWith } from "lamb";

/**
 * @template {Record<PropertyKey, any>} T
 * @template {string} U
 * @param {T} source
 * @param {U} path
 * @returns {Exclude<import("lamb").GetPath<T, U>, undefined> | []}
 */
const getPathOrEmptyArray = (source, path) => getPathIn(source, path) ?? [];

/** @type {(v: APISearchResult) => SearchResult[]} */
const transformSearchResult = compose(
  mapWith((entry) =>
    entry?.header?.hash
      ? { id: entry.header.hash, type: "block" }
      : { id: entry.txid, type: "transaction" }
  ),
  (v) => [
    ...getPathOrEmptyArray(v, "data.data.blocks"),
    ...getPathOrEmptyArray(v, "data.data.transactions"),
  ]
);

export default transformSearchResult;
