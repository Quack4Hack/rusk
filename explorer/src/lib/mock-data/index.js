export { default as apiMarketData } from "./api-market-data.json";
export { default as apiNodeLocations } from "./api-node-locations.json";
export { default as apiStats } from "./api-stats.json";
export { default as gqlBlock } from "./gql-block.json";
export { default as gqlBlocks } from "./gql-blocks.json";
export { default as gqlLatestChainInfo } from "./gql-latest-chain-info.json";
export { default as gqlSearchPossibleResults } from "./gql-search-possible-results.json";
export { default as gqlTransaction } from "./gql-transaction.json";
export { default as gqlTransactionDetails } from "./gql-transaction-details.json";
export { default as gqlTransactions } from "./gql-transactions.json";
