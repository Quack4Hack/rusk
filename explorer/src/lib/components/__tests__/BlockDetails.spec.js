import { afterAll, afterEach, describe, expect, it, vi } from "vitest";
import { cleanup, render } from "@testing-library/svelte";
import { setPathIn } from "lamb";

import { gqlBlock } from "$lib/mock-data";
import { transformBlock } from "$lib/chain-info";

import { BlockDetails } from "../";

global.ResizeObserver = vi.fn().mockImplementation(() => ({
  disconnect: vi.fn(),
  observe: vi.fn(),
  unobserve: vi.fn(),
}));

/**
 * @param {HTMLElement} container
 * @param {"next" | "prev"} which
 * @returns {HTMLAnchorElement?}
 */
const getBlockNavLink = (container, which) =>
  container.querySelector(
    `.block-details__list-anchor:nth-of-type(${which === "prev" ? "1" : "2"})`
  );

describe("Block Details", () => {
  vi.useFakeTimers();
  vi.setSystemTime(new Date(2024, 4, 20));

  const baseProps = {
    data: transformBlock(gqlBlock.block),
    error: null,
    loading: false,
  };

  afterEach(cleanup);

  afterAll(() => {
    vi.useRealTimers();
  });

  it("renders the Block Details component", () => {
    const { container } = render(BlockDetails, baseProps);

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should disable the previous block link if the prev block hash is empty or if the current height is `0`", async () => {
    const { container, rerender } = render(
      BlockDetails,
      setPathIn(baseProps, "data.header.prevblockhash", "")
    );

    expect(getBlockNavLink(container, "prev")).toHaveAttribute(
      "aria-disabled",
      "true"
    );

    rerender(baseProps);

    expect(getBlockNavLink(container, "prev")).toHaveAttribute(
      "aria-disabled",
      "false"
    );

    rerender(setPathIn(baseProps, "data.header.height", 0));

    expect(getBlockNavLink(container, "prev")).toHaveAttribute(
      "aria-disabled",
      "true"
    );
  });

  it("should disable the next block link if the next block hash is empty", () => {
    const { container } = render(
      BlockDetails,
      setPathIn(baseProps, "data.header.nextblockhash", "")
    );

    expect(getBlockNavLink(container, "next")).toHaveAttribute(
      "aria-disabled",
      "true"
    );
  });
});
