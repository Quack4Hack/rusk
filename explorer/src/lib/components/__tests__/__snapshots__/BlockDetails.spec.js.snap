// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Block Details > renders the Block Details component 1`] = `
<div>
  <div
    class="dusk-card data-card block-details"
  >
    <header
      class="data-card__header"
      slot="header"
    >
      <h1
        class="data-card__header-title"
      >
        Block Details
      </h1>
       
      <button
        class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--normal"
        type="button"
      >
        <span
          class="dusk-button__text"
        >
          Back
        </span>
      </button>
      <!--&lt;Button&gt;-->
    </header>
     
    <dl
      class="block-details__list"
    >
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The hash for the header of the block"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
        <!--&lt;Icon&gt;-->
         
        block hash
      </dt>
       
      <dd
        class="details-list__definition"
      >
        bd5c99bb720b03500e89f103fe66113ba62f2e124ed9651563f38fd15977719f
      </dd>
      <!--&lt;ListItem&gt;-->
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The height of the block indicates the length of the block chain and is increased with each additional block"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
        <!--&lt;Icon&gt;-->
         
        height
      </dt>
       
      <dd
        class="details-list__definition"
      >
        <a
          aria-disabled="false"
          class="dusk-anchor dusk-anchor-button dusk-anchor-button--variant--primary dusk-anchor-button--size--normal dusk-icon-button block-details__list-anchor"
          href="/some-base-path/blocks/block?id=07b74b35c2c7cf8f41426cd0870bafa1a2c7adee3fdd876643548096186fc4cb"
        >
          <svg
            class="dusk-icon dusk-icon--size--normal dusk-anchor-button__icon"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M20,11V13H8L13.5,18.5L12.08,19.92L4.16,12L12.08,4.08L13.5,5.5L8,11H20Z"
            />
          </svg>
          
          <!--&lt;Icon&gt;-->
           
          
          
        </a>
        <!--&lt;Anchor&gt;-->
        <!--&lt;AnchorButton&gt;-->
        <!--&lt;AppAnchorButton&gt;-->
         
        495,868
         
        <a
          aria-disabled="false"
          class="dusk-anchor dusk-anchor-button dusk-anchor-button--variant--primary dusk-anchor-button--size--normal dusk-icon-button block-details__list-anchor"
          href="/some-base-path/blocks/block?id=6011556208a85e6001bd01ccbf936486b91318a7f6cbcf7ab810adf6fae34204"
        >
          <svg
            class="dusk-icon dusk-icon--size--normal dusk-anchor-button__icon"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M4,11V13H16L10.5,18.5L11.92,19.92L19.84,12L11.92,4.08L10.5,5.5L16,11H4Z"
            />
          </svg>
          
          <!--&lt;Icon&gt;-->
           
          
          
        </a>
        <!--&lt;Anchor&gt;-->
        <!--&lt;AnchorButton&gt;-->
        <!--&lt;AppAnchorButton&gt;-->
         
      </dd>
      <!--&lt;ListItem&gt;-->
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The date and time the block was created"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
        <!--&lt;Icon&gt;-->
         
        timestamp
      </dt>
       
      <dd
        class="details-list__definition"
      >
        <time
          class="block-details__list-timestamp"
          datetime="2024-04-16T20:07:02.000Z"
          slot="definition"
        >
          Tue, 16 Apr 2024 20:07:02 GMT (last month)
           
        </time>
      </dd>
      <!--&lt;ListItem&gt;-->
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The number of transactions included in the block"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
        <!--&lt;Icon&gt;-->
         
        transactions
      </dt>
       
      <dd
        class="details-list__definition"
      >
        2
      </dd>
      <!--&lt;ListItem&gt;-->
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The total fees paid for the transactions in the block"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
        <!--&lt;Icon&gt;-->
         
        block fees paid
      </dt>
       
      <dd
        class="details-list__definition"
      >
        0.000580718
         DUSK
      </dd>
      <!--&lt;ListItem&gt;-->
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The reward allocated to the block generator"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
        <!--&lt;Icon&gt;-->
         
        block reward
      </dt>
       
      <dd
        class="details-list__definition"
      >
        16
         DUSK
      </dd>
      <!--&lt;ListItem&gt;-->
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The block gas limit"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
        <!--&lt;Icon&gt;-->
         
        block gas limit
      </dt>
       
      <dd
        class="details-list__definition"
      >
        5,000,000,000
      </dd>
      <!--&lt;ListItem&gt;-->
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The amount of gas used generating the block"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
        <!--&lt;Icon&gt;-->
         
        gas used
      </dt>
       
      <dd
        class="details-list__definition"
      >
        580,718
         
        <div
          class="dusk-progress-bar block-details__gas-used"
          role="progressbar"
        >
          <div
            class="dusk-progress-bar__filler"
            style="width: 0.01161436%"
          />
        </div>
        <!--&lt;ProgressBar&gt;-->
         
      </dd>
      <!--&lt;ListItem&gt;-->
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The average fee paid for the transactions within the block"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
        <!--&lt;Icon&gt;-->
         
        average fee paid
      </dt>
       
      <dd
        class="details-list__definition"
      >
        0.000000001
         DUSK
      </dd>
      <!--&lt;ListItem&gt;-->
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The state root hash"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
        <!--&lt;Icon&gt;-->
         
        state root hash
      </dt>
       
      <dd
        class="details-list__definition"
      >
        <span
          class="block-details__state-hash"
          slot="definition"
        >
          20bb0a677b93f084afadfd34bec3ac3feee33a020b81d9549afa2268e8543acb
        </span>
      </dd>
      <!--&lt;ListItem&gt;-->
    </dl>
    
     
  </div>
  
  <!--&lt;Card&gt;-->
  <!--&lt;DataCard&gt;-->
  <!--&lt;BlockDetails&gt;-->
</div>
`;
